### PR TITLE
Player class

### PR DESCRIPTION
**Pull request by terryzfeng**
Implement a Player class to play Numbers Royale.
Contains moves, basic player information printing, and player logic.

### CODE REVIEW

File: src/player.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+#include <string>
+
+
+class Player {
+protected:
+  std::unordered_set<unsigned int> moves_;
+  unsigned int id_;
+  unsigned int score_;
+  unsigned int last_move_;
+  unsigned int board_size_;
```
Comment (evanjserrano) [inline src/player.h:19]: Maybe I'm too embedded now that I **never** use `int` and strictly will use `(s)size_t` and `(u)intX_t`

Comment (evanjserrano) [reply src/player.h:19]: I'll choose to let it be :)

File: src/player.h [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  bool is_cpu_;
```
Comment (evanjserrano) [inline src/player.h]: I want to get rid of this variable completely, see comment at bottom of file

File: src/player.h [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  /**
+   * @brief Get if player is a CPU
+   * 
+   * @return true if player is a CPU, false otherwise
+   */
+  bool is_cpu() const {
+    return is_cpu_;
+  }
```
Comment (evanjserrano) [inline src/player.h]: Default this to be true for the Player class, and override it in the CPU class to return true
```suggestion
  virtual bool is_cpu() const {
    return false;
  }
```

Comment (evanjserrano) [reply src/player.h]: This removes `is_cpu_ completely

File: src/player.h [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+  /**
+   * @brief Get the player's moves set
+   * 
+   * @return the player's moves set
+   */
+  std::unordered_set<unsigned int> get_moves() {
+    return moves_;
+  }
```
Comment (evanjserrano) [inline src/player.h]: a big object to pass around... can we do this?
```suggestion
  const std::unordered_set<unsigned int>& get_moves() {
    return moves_;
  }
```

File: src/player.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  if (!is_valid_move(number)) {
+    return 0;
+  }
+  moves_.erase(number);
+  last_move_ = number;
+  return number;
+}
+
+std::string Player::get_moves_string() {
+  std::vector<bool> moves(board_size_);
```
Comment (evanjserrano) [inline src/player.cpp]: if this is constant size then why don't we use a `std::array` ;)

this is a nit I'm an embedded guy sorry

File: src/player.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+void rtrim(std::string& s) {
+  s.erase(s.find_last_not_of(" \n\r\t") + 1);
+} 
+
+Player::Player(bool is_cpu): is_cpu_(is_cpu) {
+  id_ = 0;
+  score_ = 0;
+  last_move_ = 0;
+  board_size_ = 0;
+}
```
Comment (evanjserrano) [inline src/player.cpp]: nit
```suggestion
Player::Player(bool is_cpu): id_(0), score_(0), last_move_(0), board_size_(0), is_cpu_(is_cpu) {}
```

File: src/player.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
+#include "player.h"
+#include <iostream>
+
+// Helper function
+/**
+ * @brief Trim trailing whitespace from a string
+ * 
+ * @param s string to trim
+ */
+void rtrim(std::string& s) {
```
Comment (evanjserrano) [inline src/player.cpp]: nit
```suggestion
static void rtrim(std::string& s) {
```

File: src/player.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  s.erase(s.find_last_not_of(" \n\r\t") + 1);
+}
```
Comment (evanjserrano) [inline src/player.cpp]: extra nit
```suggestion
namespace {

void rtrim(std::string& s) {
  s.erase(s.find_last_not_of(" \n\r\t") + 1);
} 

};
```

Comment (evanjserrano) [reply src/player.cpp]: https://medium.com/@mane.tako/understanding-unnamed-namespaces-in-c-7d41d367fd47

File: src/player.cpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+} 
+
+Player::Player(bool is_cpu): is_cpu_(is_cpu) {
+  id_ = 0;
+  score_ = 0;
+  last_move_ = 0;
+  board_size_ = 0;
+}
+
+void Player::init(unsigned int id, unsigned int board_size) {
```
Comment (evanjserrano) [inline src/player.cpp:18]: nit - do we need a constructor AND an init? do we want to call init in the ctor?

File: src/player.h (hunk 1 above)
Comment (terryzfeng) [reply src/player.h:19]: I think all the projects that I've worked on don't use int either. Always #define their own types for things, e.g. `#define CK_INT int`.

What's the (s) and (u)? signed unsigned long?

File: src/player.cpp (hunk 5 above)
Comment (terryzfeng) [reply src/player.cpp:46]: #safetyfirst

Comment (terryzfeng) [reply src/player.cpp:46]: oh nvm it doesn't work cause board_size can change?

Comment (terryzfeng) [reply src/player.cpp]: Board size can change so it won't let me define a constant size array.

File: src/player.cpp (hunk 9 above)
Comment (terryzfeng) [reply src/player.cpp:18]: Good question. I currently create players, then I pick board size. So I would have to address that somehow...

File: src/player.h (hunk 1 above)
Comment (evanjserrano) [reply src/player.h:19]: `ssize_t` is a signed `size_t`
`uint8_t` is an unsigned `int8_t`

Comment (evanjserrano) [reply src/player.h:19]: C++:
```cpp
using CK_INT = int;
```
C:
```c
typedef int CK_INT;
```

File: src/player.h (hunk 4 above)
Comment (terryzfeng) [reply src/player.h]: is it:
```cpp
std::unordered_set<unsigned int>& get_moves() const
```
or
```cpp
const std::unordered_set<unsigned int>& get_moves()
```

File: src/player.cpp (hunk 9 above)
Comment (evanjserrano) [reply src/player.cpp:18]: fine for now if you think so

File: src/player.cpp [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+   * @brief Trim trailing whitespace from a string
+   * 
+   * @param s string to trim
+   */
+  static void rtrim(std::string& s) {
+    s.erase(s.find_last_not_of(" \n\r\t") + 1);
+  } 
+};
+
+Player::Player(): id_(0), score_(0), last_move_(0), board_size_(0) {}
```
Comment (evanjserrano) [inline src/player.cpp]: Let's initialize `moves_`

File: src/player.cpp [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  return number;
+}
+
+std::string Player::get_moves_string() {
+  std::vector<bool> moves(board_size_);
+  std::string result = "";
+  // Convert to boolean vector
+  for (unsigned int number : moves_) {
+    moves[number - 1] = true;
+  }
```
Comment (evanjserrano) [inline src/player.cpp]: `moves_` vs `moves` is confusing. can you rename it to something slightly more specific/clear

Comment (evanjserrano) [reply src/player.cpp]: `moves_vec` `moves_list` idk ask chatgpt

File: src/player.cpp [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  // Build moves string
+  for (int i = 0; i < board_size_; ++i) {
```
Comment (evanjserrano) [inline src/player.cpp]: nit
```suggestion
  for (unsigned int i = 0; i < board_size_; ++i) {
```

File: src/player.cpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+   * @brief Trim trailing whitespace from a string
+   * 
+   * @param s string to trim
+   */
+  static void rtrim(std::string& s) {
+    s.erase(s.find_last_not_of(" \n\r\t") + 1);
+  } 
+};
+
+Player::Player(): id_(0), score_(0), last_move_(0), board_size_(0), moves_() {}
```
Comment (evanjserrano) [inline src/player.cpp]: I'm pretty sure this has to match the order of declarations
```cpp
  std::unordered_set<unsigned int> moves_;
  unsigned int id_;
  unsigned int score_;
  unsigned int last_move_;
  unsigned int board_size_;
```

**Review by evanjserrano — APPROVED** [review]
looks good ty :)

File: src/player.cpp (hunk 13 above)
Comment (terryzfeng) [reply src/player.cpp]: :0 I learned something new

**Pull request merged**